bootstrap.rkt

```diff
@@ -0,0 +1,476 @@
+                                                                                       ,rhs-head]
+                                                                                      [_ (failure-cc)])
+                                                                                    stx))))))))))]
+                                                           [_
+                                                            (pure (generate-quasiquote
+                                                                    (failure-cc)
+                                                                    stx))])))])]
+                                             (flet [fancy-cases (cases inner-cases)
+                                                    (generate-syntax-case syntax-case cases ()
+                                                      [()
+                                                       (pure inner-cases)]
+                                                      [(,case ,cases ...)
+                                                       (>>= (fancy-case stx-name case)
+                                                         (lambda (inner-case)
+                                                           (fancy-cases cases
+                                                             (generate-quasiquote
+                                                               (let [failure-cc
+                                                                     (lambda ()
+                                                                       ,inner-case)]
+                                                                 ,inner-cases)
+                                                               stx))))])]
+                                               (>>= (fancy-cases cases
+                                                      (generate-quasiquote
+                                                        (failure-cc)
+                                                        stx))
+                                                 (lambda (outer-cases)
+                                                   (pure (generate-quasiquote
+                                                           (let [,stx-name ,stx-expr]
+                                                             (let [failure-cc
+                                                                   (lambda ()
+                                                                     (syntax-error (list-syntax
+                                                                                     ('"syntax-case: the input"
+                                                                                      ,stx-name
+                                                                                      '"does not match any of the following patterns"
+                                                                                      ',(map car cases))
```

Comment:
I spent some effort on the quality of my error messages. Most of them are for when `syntax-case` itself is misused, but I'm especially proud of this one: if you define a macro using fancy-syntax-case, the callers of _that_ macro also get a nice error message if they misuse that macro!